bigtable/cmd/cbt/cbt.go

```diff
@@ -988,49 +994,91 @@ func doListClusters(ctx context.Context, args ...string) {
 	tw.Flush()
 }
 
-func doLookup(ctx context.Context, args ...string) {
-	if len(args) < 2 {
-		log.Fatalf("usage: cbt lookup <table> <row> [columns=<family:qualifier>...] [cells-per-column=<n>] " +
-			"[app-profile=<app profile id>]")
-	}
+func getDataFilter(
+	parsed map[string]string, filters ...bigtable.Filter,
+) (bigtable.ReadOption, bool, error) {
+	var err error
 
-	parsed, err := parseArgs(args[2:], []string{"columns", "cells-per-column", "app-profile"})
-	if err != nil {
-		log.Fatal(err)
-	}
-	var opts []bigtable.ReadOption
-	var filters []bigtable.Filter
 	if cellsPerColumn := parsed["cells-per-column"]; cellsPerColumn != "" {
 		n, err := strconv.Atoi(cellsPerColumn)
 		if err != nil {
-			log.Fatalf("Bad number of cells per column %q: %v", cellsPerColumn, err)
+			return nil, false, fmt.Errorf(
+				"Bad number of cells per column %q: %v",
+				cellsPerColumn, err)
 		}
 		filters = append(filters, bigtable.LatestNFilter(n))
 	}
+
+	keysOnly := false
+	if keysOnlyS := parsed["keys-only"]; keysOnlyS != "" {
+		keysOnly, err = strconv.ParseBool(keysOnlyS)
+		if err != nil {
+			return nil, false, fmt.Errorf(
+				"Bad value for keys-only: %v", keysOnlyS)
+		}
+		if keysOnly {
+			filters = append(filters, bigtable.StripValueFilter())
+		}
+	}
+
 	if columns := parsed["columns"]; columns != "" {
 		columnFilters, err := parseColumnsFilter(columns)
 		if err != nil {
-			log.Fatal(err)
+			return nil, keysOnly, err
 		}
 		filters = append(filters, columnFilters)
 	}
 
-	if len(filters) > 1 {
-		opts = append(opts, bigtable.RowFilter(bigtable.ChainFilters(filters...)))
-	} else if len(filters) == 1 {
-		opts = append(opts, bigtable.RowFilter(filters[0]))
+	var filter bigtable.Filter = nil
+	var option bigtable.ReadOption
+	if len(filters) > 0 {
+		if len(filters) > 1 {
+			filter = bigtable.ChainFilters(filters...)
+		} else if len(filters) == 1 {
+			filter = filters[0]
+		}
+		option = bigtable.RowFilter(filter)
+	}
+
+	return option, keysOnly, nil
+}
+
+func doLookup(ctx context.Context, args ...string) {
+	if len(args) < 2 {
+		log.Fatalf("usage: cbt lookup <table> <row> [columns=<family:qualifier>...] [cells-per-column=<n>] " +
+			"[app-profile=<app profile id>]")
+	}
+
+	parsed, err := parseArgs(
+		args[2:], []string{
+			"columns", "cells-per-column", "app-profile", "keys-only"},
+	)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	opt, keysOnly, err := getDataFilter(parsed)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var opts []bigtable.ReadOption
+	if opt != nil {
+		opts = append(opts, opt)
 	}
 
 	table, row := args[0], args[1]
-	tbl := getClient(bigtable.ClientConfig{AppProfile: parsed["app-profile"]}).Open(table)
+	tbl := getTable(
+		bigtable.ClientConfig{AppProfile: parsed["app-profile"]},
+		table)
 	r, err := tbl.ReadRow(ctx, row, opts...)
```

Comment:
It seems that opts can only have a single entry in it, if this is the case, what is the reason for unpacking it? Is this to avoid having to nil-check here?

Reply:
`opts` can have 0 or 1 entries in it.  So the unpacking handles both cases.  Of course, this is the original code, which I tried not to disturb (aside from factoring out the filtering logic, which I now kind of regret :smiley: ).

Reply:
> factoring out the filtering logic, which I now kind of regret

Well, no. Factoring out the filtering code was a great idea :smiley:, because it was repeated for read and lookup.